[AppSSO] Adjust life cycle of SOAuthorization to avoid premature cancellation

brentfulgham opened the pull request:
#### bc90c50c6ba84974bbc861bcbb406bbcd35e845f
<pre>
[AppSSO] Adjust life cycle of SOAuthorization to avoid premature cancellation
<a href="https://bugs.webkit.org/show_bug.cgi?id=244364">https://bugs.webkit.org/show_bug.cgi?id=244364</a>
&lt;rdar://98413693&gt;

Reviewed by J Pascoe.

This patch modifies the SOAuthorizationSession logic to own the underlying SOAuthorization
Cocoa object. This simplifies the life cycle of the object, and helps clarify which
object is associated with a given load.

Originally we had thought that letting the SOAuthorizationController hold the SOAuthorization
would allow it to be reused in some cases, but in practice no such feature was needed, and
there is little performance gain from retaining a single object for a set of navigations.

Tested by existing SOAuthorization API tests.

* Source/WebKit/UIProcess/Cocoa/SOAuthorization/NavigationSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/NavigationSOAuthorizationSession.mm:
(WebKit::NavigationSOAuthorizationSession::NavigationSOAuthorizationSession):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/PopUpSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/PopUpSOAuthorizationSession.mm:
(WebKit::PopUpSOAuthorizationSession::create):
(WebKit::PopUpSOAuthorizationSession::PopUpSOAuthorizationSession):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/RedirectSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/RedirectSOAuthorizationSession.mm:
(WebKit::RedirectSOAuthorizationSession::create):
(WebKit::RedirectSOAuthorizationSession::RedirectSOAuthorizationSession):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationCoordinator.mm:
(WebKit::SOAuthorizationCoordinator::SOAuthorizationCoordinator):
(WebKit::SOAuthorizationCoordinator::tryAuthorize):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationSession.mm:
(WebKit::SOAuthorizationSession::SOAuthorizationSession):
(WebKit::SOAuthorizationSession::setSOAuthorizationDelegate):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.mm:
(WebKit::SubFrameSOAuthorizationSession::create):
(WebKit::SubFrameSOAuthorizationSession::SubFrameSOAuthorizationSession):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/WKSOAuthorizationDelegate.mm:
(-[WKSOAuthorizationDelegate setSession:]):

Canonical link: <a href="https://commits.webkit.org/253929@main">https://commits.webkit.org/253929@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c55c4361a204ad331e7f80ecc4cb10611d5fef3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87201 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31288 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18029 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96231 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29652 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25882 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79317 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91213 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92817 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23955 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74006 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23844 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78931 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79176 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66858 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27377 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13013 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27324 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14028 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2761 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29009 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36884 "Found 2 new test failures: http/wpt/crypto/aes-ctr-crash.any.worker.html, http/wpt/crypto/rsa-oaep-crash.any.worker.html") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28948 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33305 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/SOAuthorization/WKSOAuthorizationDelegate.mm:
```diff
@@ -144,9 +144,19 @@ - (void)setSession:(RefPtr<WebKit::SOAuthorizationSession>&&)session
+        _session->setSOAuthorizationDelegate(self);
         _session->shouldStart();
+    } else
+        WTFLogAlways("WTF? No Session?");
```
Whoops!

pascoej reviewed (approved):
r+. I have one question that I'm going to ask offline.

brentfulgham committed:
[AppSSO] Adjust life cycle of SOAuthorization to avoid premature cancellation
https://bugs.webkit.org/show_bug.cgi?id=244364
<rdar://98413693>

Reviewed by J Pascoe.

This patch modifies the SOAuthorizationSession logic to own the underlying SOAuthorization
Cocoa object. This simplifies the life cycle of the object, and helps clarify which
object is associated with a given load.

Originally we had thought that letting the SOAuthorizationController hold the SOAuthorization
would allow it to be reused in some cases, but in practice no such feature was needed, and
there is little performance gain from retaining a single object for a set of navigations.

Tested by existing SOAuthorization API tests.

* Source/WebKit/UIProcess/Cocoa/SOAuthorization/NavigationSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/NavigationSOAuthorizationSession.mm:
(WebKit::NavigationSOAuthorizationSession::NavigationSOAuthorizationSession):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/PopUpSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/PopUpSOAuthorizationSession.mm:
(WebKit::PopUpSOAuthorizationSession::create):
(WebKit::PopUpSOAuthorizationSession::PopUpSOAuthorizationSession):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/RedirectSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/RedirectSOAuthorizationSession.mm:
(WebKit::RedirectSOAuthorizationSession::create):
(WebKit::RedirectSOAuthorizationSession::RedirectSOAuthorizationSession):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationCoordinator.mm:
(WebKit::SOAuthorizationCoordinator::SOAuthorizationCoordinator):
(WebKit::SOAuthorizationCoordinator::tryAuthorize):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationSession.mm:
(WebKit::SOAuthorizationSession::SOAuthorizationSession):
(WebKit::SOAuthorizationSession::setSOAuthorizationDelegate):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.mm:
(WebKit::SubFrameSOAuthorizationSession::create):
(WebKit::SubFrameSOAuthorizationSession::SubFrameSOAuthorizationSession):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/WKSOAuthorizationDelegate.mm:
(-[WKSOAuthorizationDelegate setSession:]):

Canonical link: https://commits.webkit.org/253929@main
